ibis/sql/tests/test_compiler.py

```diff
@@ -2496,3 +2498,27 @@ def test_table_drop_with_filter():
     ON t0.`b` = t1.`b`
 WHERE t0.`a` < 1.0"""
     assert result == expected
+
+
+def test_large_compile():
+    """
+    Tests that compiling a large expression tree finishes
+    within a reasonable amount of time
+    """
+    num_columns = 20
+    num_joins = 7
+
+    names = [f"col_{i}" for i in range(num_columns)]
+    schema = ibis.Schema(names, ['string'] * num_columns)
+    ibis_client = bq.BigQueryClient.__new__(bq.BigQueryClient)
```

Comment:
Ok, I think I can get this to work 

```
In [42]: class MockBigQueryClient(BigQueryClient): 
    ...:     def __init__(self): 
    ...:         pass 
    ...:                                                                                                                                                                                                    

In [43]: client = MockBigQueryClient()                                                                                                                                                                      

In [44]: table = TableExpr(ops.SQLQueryResult("select * from t", schema, client))                                                                                                                           

In [45]: table.compile()                                                                                                                                                                                    
Out[45]: 'SELECT *\nFROM (\n  select * from t\n) t0'

In [46]: for _ in range(num_joins): 
    ...:     table = table.mutate(dummy=ibis.literal("")) 
    ...:     table = table.left_join(table, ["dummy"])[[table]] 
    ...:                                                                                                                                                                                                    

In [47]: table.compile()                                                                                                                                                                                    
Out[47]: "WITH t0 AS (\n  select * from t\n),\nt1 AS (\n  SELECT *, '' AS `dummy`\n  FROM t0\n),\nt2 AS (\n  SELECT t1.*\n  FROM t1\n    LEFT OUTER JOIN t1 t16\n      ON t1.`dummy` = t16.`dummy`\n),\nt3 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t2\n),\nt4 AS (\n  SELECT t3.*\n  FROM t3\n    LEFT OUTER JOIN t3 t16\n      ON t3.`dummy` = t16.`dummy`\n),\nt5 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t4\n),\nt6 AS (\n  SELECT t5.*\n  FROM t5\n    LEFT OUTER JOIN t5 t16\n      ON t5.`dummy` = t16.`dummy`\n),\nt7 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t6\n),\nt8 AS (\n  SELECT t7.*\n  FROM t7\n    LEFT OUTER JOIN t7 t16\n      ON t7.`dummy` = t16.`dummy`\n),\nt9 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t8\n),\nt10 AS (\n  SELECT t9.*\n  FROM t9\n    LEFT OUTER JOIN t9 t16\n      ON t9.`dummy` = t16.`dummy`\n),\nt11 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t10\n),\nt12 AS (\n  SELECT t11.*\n  FROM t11\n    LEFT OUTER JOIN t11 t16\n      ON t11.`dummy` = t16.`dummy`\n),\nt13 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t12\n)\nSELECT t13.*\nFROM t13\n  LEFT OUTER JOIN t13 t14\n    ON t13.`dummy` = t14.`dummy`"

In [48]: %time table.compile()                                                                                                                                                                              
CPU times: user 38.8 s, sys: 245 ms, total: 39.1 s
Wall time: 40.2 s
Out[48]: "WITH t0 AS (\n  select * from t\n),\nt1 AS (\n  SELECT *, '' AS `dummy`\n  FROM t0\n),\nt2 AS (\n  SELECT t1.*\n  FROM t1\n    LEFT OUTER JOIN t1 t16\n      ON t1.`dummy` = t16.`dummy`\n),\nt3 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t2\n),\nt4 AS (\n  SELECT t3.*\n  FROM t3\n    LEFT OUTER JOIN t3 t16\n      ON t3.`dummy` = t16.`dummy`\n),\nt5 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t4\n),\nt6 AS (\n  SELECT t5.*\n  FROM t5\n    LEFT OUTER JOIN t5 t16\n      ON t5.`dummy` = t16.`dummy`\n),\nt7 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t6\n),\nt8 AS (\n  SELECT t7.*\n  FROM t7\n    LEFT OUTER JOIN t7 t16\n      ON t7.`dummy` = t16.`dummy`\n),\nt9 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t8\n),\nt10 AS (\n  SELECT t9.*\n  FROM t9\n    LEFT OUTER JOIN t9 t16\n      ON t9.`dummy` = t16.`dummy`\n),\nt11 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t10\n),\nt12 AS (\n  SELECT t11.*\n  FROM t11\n    LEFT OUTER JOIN t11 t16\n      ON t11.`dummy` = t16.`dummy`\n),\nt13 AS (\n  SELECT `col_0`, `col_1`, `col_2`, `col_3`, `col_4`, `col_5`, `col_6`, `col_7`,\n         `col_8`, `col_9`, `col_10`, `col_11`, `col_12`, `col_13`,\n         `col_14`, `col_15`, `col_16`, `col_17`, `col_18`, `col_19`,\n         '' AS `dummy`\n  FROM t12\n)\nSELECT t13.*\nFROM t13\n  LEFT OUTER JOIN t13 t14\n    ON t13.`dummy` = t14.`dummy`"

Reply:
I think this is a little clearer than `bq.BigQueryClient.__new__(bq.BigQueryClient)`